src/opencl_solarwinds_fmt_plug.c

```diff
@@ -0,0 +1,368 @@
+/*
+ * JtR OpenCL format to crack SolarWinds Orion hashes.
+ *
+ * This software is Copyright (c) 2018, Dhiru Kholia <dhiru at openwall.com>,
+ * and it is hereby released to the general public under the following terms:
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * The OpenCL boilerplate code is borrowed from other OpenCL formats.
+ */
+
+#ifdef HAVE_OPENCL
+
+#if FMT_EXTERNS_H
+extern struct fmt_main fmt_opencl_solarwinds;
+#elif FMT_REGISTERS_H
+john_register_one(&fmt_opencl_solarwinds);
+#else
+
+#include <stdint.h>
+#include <string.h>
+
+#include "aes.h"
+#include "arch.h"
+#include "formats.h"
+#include "common.h"
+#include "solarwinds_common.h"
+#include "options.h"
+#include "jumbo.h"
+#include "opencl_common.h"
+#include "misc.h"
+#define OUTLEN                  1024
+#define PLAINTEXT_LENGTH        28
+#include "opencl_pbkdf2_hmac_sha1.h"
+
+#define FORMAT_LABEL            "solarwinds-opencl"
+#define OCL_ALGORITHM_NAME      "PBKDF2-SHA1 OpenCL"
+#define ALGORITHM_NAME          OCL_ALGORITHM_NAME
+#define BENCHMARK_COMMENT       ""
+#define BENCHMARK_LENGTH        -1
+#define MIN_KEYS_PER_CRYPT      1
+#define MAX_KEYS_PER_CRYPT      1
+
+#define BINARY_SIZE             64
+#define BINARY_ALIGN            MEM_ALIGN_WORD
+#define SALT_SIZE               sizeof(*cur_salt)
+#define SALT_ALIGN              MEM_ALIGN_WORD
+
+/* This handles all widths */
+#define GETPOS(i, index)        (((index) % ocl_v_width) * 4 + ((i) & ~3U) * ocl_v_width + (((i) & 3) ^ 3) + ((index) / ocl_v_width) * 64 * ocl_v_width)
+
+static struct custom_salt *cur_salt;
+
+typedef struct {
+	unsigned int key[((OUTLEN + 19) / 20) * 20 / sizeof(uint32_t)];
+} solarwinds_out;
+
+typedef struct {
+	unsigned char hash[12];
+} solarwinds_final_out;
+
+typedef struct {
+	unsigned int  length;
+	unsigned int  outlen;
+	unsigned int  iterations;
+	unsigned char salt[8];
+} solarwinds_salt;
+
+static size_t key_buf_size, outsize, final_outsize;
+static unsigned int *inbuffer;
+static solarwinds_out *output;
+static solarwinds_final_out *final_output;
+static solarwinds_salt currentsalt;
+static cl_mem mem_in, mem_out, mem_salt, mem_state, mem_final_out;
+static size_t key_buf_size;
+static int new_keys;
+static struct fmt_main *self;
+static cl_kernel pbkdf2_init, pbkdf2_loop, solarwinds_loop, solarwinds_final;
+
+/*
+ * HASH_LOOPS is ideally made by factors of (iteration count - 1) and should
+ * be chosen for a kernel duration of not more than 200 ms
+ */
+#define HASH_LOOPS              (3 * 271)
+#define ITERATIONS              100000 /* Just for auto tune */
```

Comment:
Drop this `#define ITERATIONS` line, it's an artefact from something else. Or maybe better: Set it to 1000 if that is the fixed value, and pass it to the kernel (in line 193) for possibly better optimization by the runtime.

Reply:
This seems to be used in line 215 (`autotune_run(self, 2 * (ITERATIONS - 1) + 4, 0, 200)`).

Reply:
OK but then it has to be correct or auto-tune will end up incorrect by three orders of magnitude. That comment "Just for auto tune" sounds like it's not important - it is!

So set it to 1000 and while at it, add it to `build_opts` as well, like I said may end up in a better optimized loop kernel.

Reply:
Did you watch an auto-tune with `-v:5`? Do so and ensure it looks correct.

Reply:
BTW where did you get `3 * 271` from? It doesn't look like a good figure for splitting 1000 iterations into several calls. Something like 333 would match the comment above it.

Reply:
It came from a copy-paste operation ;(. I need to sit down and read this auto-tune stuff one day.

Reply:
All these changes should be in place now. Thanks!

Reply:
I think `HASH_LOOPS` should be just `333` as opposed to `3 * 333`. Current code is effectively a non-split kernel: It only runs the loop kernel once (using 333 it would be three times). Current figures run fine on these high-end cards but on a weaker device it will put a limit on reaching optimal GWS.

Reply:
Ack. I have pushed a new commit now.